File: src/api/assignments.rs
use axum::{
    extract::{Path, State},
    Json,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::{
    auth::middleware::{require_admin, AuthUser},
    error::{AppError, Result},
    models::{assignment::Assignment, device::Device},
    AppState,
};

pub async fn list_assignments(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path(device_id): Path<Uuid>,
) -> Result<Json<Vec<Assignment>>> {
    let device = Device::find_by_id(device_id, &state.db)
        .await?
        .ok_or(AppError::NotFound)?;
    if device.org_id != Some(claims.org_id) {
        return Err(AppError::NotFound);
    }

    let assignments = Assignment::list_for_device(device_id, &state.db).await?;
    Ok(Json(assignments))
}

#[derive(Deserialize)]
pub struct AssignRequest {
    pub user_id: Uuid,
}

pub async fn create_assignment(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path(device_id): Path<Uuid>,
    Json(body): Json<AssignRequest>,
) -> Result<Json<Assignment>> {
    require_admin(&claims)?;

    let device = Device::find_by_id(device_id, &state.db)
        .await?
        .ok_or(AppError::NotFound)?;
    if device.org_id != Some(claims.org_id) {
        return Err(AppError::NotFound);
    }

    let assignment = Assignment::create(device_id, body.user_id, claims.sub, &state.db).await?;
    Ok(Json(assignment))
}

pub async fn delete_assignment(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path((device_id, user_id)): Path<(Uuid, Uuid)>,
) -> Result<Json<serde_json::Value>> {
    require_admin(&claims)?;

    let device = Device::find_by_id(device_id, &state.db)
        .await?
        .ok_or(AppError::NotFound)?;
    if device.org_id != Some(claims.org_id) {
        return Err(AppError::NotFound);
    }

    let deleted = Assignment::delete(device_id, user_id, &state.db).await?;
    if !deleted {
        return Err(AppError::NotFound);
    }

    Ok(Json(serde_json::json!({ "deleted": true })))
}

File: src/api/assignments.ts
import { Hono } from 'hono'
import type { AppEnv } from '../types.js'
import { AppError } from '../error.js'
import { authMiddleware, requireAdmin } from '../auth/middleware.js'

const assignments = new Hono<AppEnv>()

assignments.use(authMiddleware)

assignments.get('/:id/assignments', async (c) => {
  const { state, user } = c.var
  const deviceId = c.req.param('id')

  const [device] = await state.db`SELECT * FROM devices WHERE id = ${deviceId}`
  if (!device || device.org_id !== user.org_id) throw AppError.notFound()

  const rows = await state.db`
    SELECT id, device_id, user_id, assigned_at, assigned_by
    FROM device_assignments WHERE device_id = ${deviceId}
  `
  return c.json(rows)
})

assignments.post('/:id/assignments', async (c) => {
  const { state, user } = c.var
  requireAdmin(user)
  const deviceId = c.req.param('id')

  const [device] = await state.db`SELECT * FROM devices WHERE id = ${deviceId}`
  if (!device || device.org_id !== user.org_id) throw AppError.notFound()

  const { user_id } = await c.req.json<{ user_id: string }>()

  const [assignment] = await state.db`
    INSERT INTO device_assignments (device_id, user_id, assigned_by)
    VALUES (${deviceId}, ${user_id}, ${user.sub})
    ON CONFLICT (device_id, user_id) DO UPDATE SET assigned_by = ${user.sub}
    RETURNING id, device_id, user_id, assigned_at, assigned_by
  `
  return c.json(assignment)
})

assignments.delete('/:id/assignments/:userId', async (c) => {
  const { state, user } = c.var
  requireAdmin(user)
  const deviceId = c.req.param('id')
  const userId = c.req.param('userId')

  const [device] = await state.db`SELECT * FROM devices WHERE id = ${deviceId}`
  if (!device || device.org_id !== user.org_id) throw AppError.notFound()

  const result = await state.db`
    DELETE FROM device_assignments WHERE device_id = ${deviceId} AND user_id = ${userId}
  `
  if (result.count === 0) throw AppError.notFound()

  return c.json({ deleted: true })
})

export default assignments
